File: bigquerydatapolicy/v2/python/google-cloud-bigquery-datapolicies/data_policy_service_client_iam_permissions_test.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# [START bigquerydatapolicy_v2_datapolicyservice_iam_permissions_test]
from google.api_core import exceptions
from google.cloud import bigquery_datapolicies_v2
from google.iam.v1 import iam_policy_pb2

client = bigquery_datapolicies_v2.DataPolicyServiceClient()


def test_data_policy_iam_permissions(
    project_id: str,
    location: str,
    data_policy_id: str,
    permissions: list[str],
) -> None:
    """Test IAM permissions for a data policy.
    Tests whether the caller has the specified IAM permissions for a data
    policy resource.

    Args:
        project_id: The Google Cloud project ID.
        location: The geographic location of the data policy (for example, 'us').
        data_policy_id: The ID of the data policy to test permissions for.
        permissions: A list of permissions to test (for example, ['bigquery.dataPolicies.get']).
    """

    resource_name = client.data_policy_path(
        project=project_id,
        location=location,
        data_policy=data_policy_id,
    )

    request = iam_policy_pb2.TestIamPermissionsRequest(
        resource=resource_name,
        permissions=permissions,
    )

    try:
        response = client.test_iam_permissions(request=request)

        print(f"Tested permissions for data policy: {resource_name}")
        if response.permissions:
            print(
                f"Caller has the following permissions: {', '.join(response.permissions)}"
            )
        else:
            print("Caller has NONE of the requested permissions.")

    except exceptions.NotFound:
        print(f"Error: Data policy '{resource_name}' not found.")
    except exceptions.GoogleAPIError as e:
        print(f"An API error occurred: {e}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")


# [END bigquerydatapolicy_v2_datapolicyservice_iam_permissions_test]

File: bigquerydatapolicy/v2/nodejs/google-cloud-bigquery-datapolicies/data-policy-service-client-iam-permissions-test.js
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

// [START bigquerydatapolicy_v2_datapolicyservice_iam_permissions_test]
const {DataPolicyServiceClient} =
  require('@google-cloud/bigquery-datapolicies').v2;
const {status} = require('@grpc/grpc-js');

const client = new DataPolicyServiceClient();

/**
 * Test IAM permissions for a data policy.
 * Tests whether the caller has the specified IAM permissions for a data
 * policy resource.
 *
 * @param {string} projectId The Google Cloud project ID (for example, 'example-project-id').
 * @param {string} location The geographic location of the data policy (for example, 'us').
 * @param {string} dataPolicyId The ID of the data policy to test permissions for.
 * @param {string[]} permissions A list of permissions to test (for example, ['bigquery.dataPolicies.get']).
 */
async function testDataPolicyIamPermissions(
  projectId,
  location,
  dataPolicyId,
  permissions,
) {
  const resourceName = client.dataPolicyPath(projectId, location, dataPolicyId);

  // Ensure permissions is an array (helper for CLI runner)
  const permissionsArray = Array.isArray(permissions)
    ? permissions
    : [permissions];

  const request = {
    resource: resourceName,
    permissions: permissionsArray,
  };

  try {
    const [response] = await client.testIamPermissions(request);

    console.log(`Tested permissions for data policy: ${resourceName}`);
    if (response.permissions && response.permissions.length > 0) {
      console.log(
        `Caller has the following permissions: ${response.permissions.join(', ')}`,
      );
    } else {
      console.log('Caller has NONE of the requested permissions.');
    }
  } catch (err) {
    if (err.code === status.NOT_FOUND) {
      console.error(`Error: Data policy '${resourceName}' not found.`);
    } else {
      console.error('An API error occurred:', err.message);
    }
  }
}
// [END bigquerydatapolicy_v2_datapolicyservice_iam_permissions_test]

module.exports = {
  testDataPolicyIamPermissions,
};
